cli/replicaset/cmd/upgrade.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	// Create a map for fast lookup of replicasets by alias
+	replicasetMap := make(map[string]replicaset.Replicaset)
+	for _, rs := range replicasets.Replicasets {
+		replicasetMap[rs.Alias] = rs
+	}
+
+	var chosenReplicasets []replicaset.Replicaset
+	for _, alias := range ChosenReplicasetAliases {
+		rs, exists := replicasetMap[alias]
+		if !exists {
+			return nil, fmt.Errorf("replicaset with alias %q doesn't exist", alias)
+		}
+		chosenReplicasets = append(chosenReplicasets, rs)
+	}
+
+	return chosenReplicasets, nil
+}
+
+func Upgrade(discoveryCtx DiscoveryCtx) error {
+	replicasets, err := GetReplicasets(discoveryCtx)
+	if err != nil {
+		// This may be a single-instance application without Tarantool-3 config
+		// or instances.yml file.
+		if len(discoveryCtx.RunningCtx.Instances) == 1 {
+			// Create a dummy replicaset
+			var replicasetList []replicaset.Replicaset
+			var dummyReplicaset replicaset.Replicaset
+			var instance replicaset.Instance
+
+			instance.InstanceCtx = discoveryCtx.RunningCtx.Instances[0]
+			instance.Alias = running.GetAppInstanceName(instance.InstanceCtx)
+			instance.InstanceCtxFound = true
+
+			dummyReplicaset.Alias = instance.Alias
+			dummyReplicaset.Instances = append(dummyReplicaset.Instances, instance)
+			replicasetList = append(replicasetList, dummyReplicaset)
+
+			return internalUpgrade(replicasetList)
+		}
+		return err
+	}
+
+	replicasets = FillAliases(replicasets)
+	replicasetsToUpgrade, err := FilterReplicasetsByAliases(replicasets)
+	if err != nil {
+		return err
+	}
+
+	return internalUpgrade(replicasetsToUpgrade)
+}
+
+func internalUpgrade(replicasets []replicaset.Replicaset) error {
+	for _, replicaset := range replicasets {
+		err := upgradeReplicaset(replicaset)
+		if err != nil {
+			fmt.Printf("• %s: error\n", replicaset.Alias)
+			return fmt.Errorf("replicaset %s: %w", replicaset.Alias, err)
+		}
+		fmt.Printf("• %s: ok\n", replicaset.Alias)
+	}
+	return nil
+}
+
+func getInstanceConnector(instance replicaset.Instance) (connector.Connector, error) {
+	run := instance.InstanceCtx
+	fullInstanceName := running.GetAppInstanceName(run)
+	if fullInstanceName == "" {
+		fullInstanceName = instance.Alias
+	}
+	if fullInstanceName == "" {
+		fullInstanceName = "unknown"
+	}
+
+	// Try to connect via unix socket
+	conn, err := connector.Connect(connector.ConnectOpts{
+		Network: "unix",
+		Address: run.ConsoleSocket,
+	})
+
+	if err != nil {
+		// try to connect via TCP [experimental]
+		conn, err = connector.Connect(connector.ConnectOpts{
+			Network:  "tcp",
+			Address:  instance.URI,
+			Username: "client", // should be opt
+			Password: "secret", // should be opt
+		})
+		if err != nil {
+			return nil, fmt.Errorf("instance %s failed to connect via both TCP "+
+				"and UNIX socket [%s]: %w", fullInstanceName, instance.URI, err)
+		}
+	}
+	return conn, nil
+}
+
+func СollectRWROInfo(replicaset replicaset.Replicaset, master **InstanceMeta,
+	replicas *[]InstanceMeta) error {
```

Comment:
Please, avoid pointers to pointers if it is not really required in Go.
```suggestion
func СollectRWROInfo(replicaset replicaset.Replicaset) (InstanceMeta, []InstanceMeta, error) {
```

Reply:
Replaced with `func СollectRWROInfo(replicaset replicaset.Replicaset) (InstanceMeta, []InstanceMeta, error)`